liana/method/sp/_misty.py

```diff
@@ -0,0 +1,471 @@
+
+import numpy as np
+from scipy.spatial import cKDTree
+from scipy.sparse import identity, issparse
+import logging
+from sklearn.ensemble import RandomForestRegressor
+from sklearn.linear_model import Ridge
+from sklearn.model_selection import KFold
+import anndata as ad
+import pandas as pd
+import squidpy as sq
+
+
+def _gauss_weight(distance_mtx, l):
+    return np.exp(-distance_mtx**2 / l**2)
+
+
+def _exponential_weight(distance_mtx, l):
+    return np.exp(-distance_mtx / l)
+
+
+def _get_distance_weights(adata, bandwidth, kernel="gaussian", add_self=True, spatial_key="spatial", zoi=0):
+    kdtree = cKDTree(adata.obsm[spatial_key])
+    if kernel == "gaussian":
+        max_dist = 3*bandwidth
+    elif kernel == "exponential":
+        max_dist = 5*bandwidth
+    sdm = kdtree.sparse_distance_matrix(kdtree, max_distance=max_dist, output_type="coo_matrix")
+    sdm = sdm.tocsr()
+    sdm.data[sdm.data < zoi] = np.inf
+    if kernel == "gaussian":
+        sdm.data = _gauss_weight(sdm.data, bandwidth)
+    elif kernel == "exponential":
+        sdm.data = _exponential_weight(sdm.data, bandwidth)
+    if not add_self:
+        sdm -= identity(n=sdm.shape[0], format="csr", dtype=sdm.dtype)
+    return(sdm)
+
+
+def _get_neighbors(adata, juxta_cutoff=np.inf, add_self=True, spatial_key="spatial"):
+    neighbors, dists = sq.gr.spatial_neighbors(adata, coord_type="generic", 
+                                               copy=True, delaunay=True, spatial_key=spatial_key)
+    neighbors[dists > juxta_cutoff] = 0
+    if add_self:
+        neighbors += identity(neighbors.shape[0]) # add self for visium (not for the other assay though)
+    return(neighbors)
+
+
+def _check_targets(adata_target, targets):
+    if targets is not None:
+        missing_targets = set(targets).difference(set(adata_target.var_names))
+        if len(missing_targets) > 0:
+            logging.warning(f"Missing target features: {missing_targets}.")
+        targets = list(set(targets).difference(missing_targets))
+    else:
+        targets = adata_target.var_names.tolist()
+    return(targets)
+
+
+def _check_predictors(adata_source, predictors):
```

Comment:
same as `_check_targets`

Reply:
Now there is a single function called `_check_features`.